### PR TITLE
Ensure that we don't execute Initialize and compilation start actions…

### LINKED ISSUE #15184: Don't execute Initialize and compilation start actions for analyzers with no enabled rules
We end up running Initialize and compilation start actions for suppressed analyzers when computing analyzer telemetry.

This is causing https://github.com/dotnet/roslyn-project-system/issues/578.

### PR DESCRIPTION
**Pull request by mavasani**
… for analyzers with no enabled rules

Fixes #15184

### CODE REVIEW

**Comment by mavasani** [comment]
**Customer scenario** - We end up executing Initialize and CompilationStartActions for analyzers which have no enabled rules and hence can never report diagnostics. If the analyzer throws in these methods, we will start seeing exception diagnostics from it, even though user explicitly suppressed it.

**Bug** - https://github.com/dotnet/roslyn/issues/15184 and https://github.com/dotnet/roslyn-project-system/issues/578

**Workarounds** - None.

**Risk** - This shouldn't have any risk as we are just checking if analyzer is suppressed or not when reporting analyzer action counts telemetry.

**Performance impact** - None. In fact, it might improve performance a bit as we don't execute these unnecessary analyzer callbacks.

**Is this a regression from a previous update?** - Yes, this probably regressed in Dev14 U3 when we added analyzer telemetry APIs and IDE started invoking them.

**Comment by mavasani** [comment]
Tagging @dotnet/roslyn-analysis @dotnet/project-system for review

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs [hunk 1]
```diff
@@ -129,7 +129,7 @@ private CompilationWithAnalyzers(Compilation compilation, ImmutableArray<Diagnos
             _cancellationToken = cancellationToken;
 
             _compilationData = new CompilationData(_compilation);
-            _analysisState = new AnalysisState(analyzers, _compilationData);
+            _analysisState = new AnalysisState(analyzers, _compilationData, _compilation.Options);
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: compilation data itself seems not hold onto compilation but it holds onto suppression state which internally hold onto compilation. so, looks like you don't need to add one more parameter.

**Comment by heejaechang** [comment]
👍

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 2]
```diff
@@ -1298,9 +1298,14 @@ protected bool IsGeneratedCode(SyntaxTree tree)
 
         protected bool DoNotAnalyzeGeneratedCode => _doNotAnalyzeGeneratedCode;
 
-        internal async Task<AnalyzerActionCounts> GetAnalyzerActionCountsAsync(DiagnosticAnalyzer analyzer, CancellationToken cancellationToken)
+        internal async Task<AnalyzerActionCounts> GetAnalyzerActionCountsAsync(DiagnosticAnalyzer analyzer, CompilationOptions compilationOptions, CancellationToken cancellationToken)
         {
             var executor = analyzerExecutor.WithCancellationToken(cancellationToken);
+            if (IsDiagnosticAnalyzerSuppressed(analyzer, compilationOptions, analyzerManager, executor))
+            {
+                return new AnalyzerActionCounts(null);
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Sure.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: CompilationData is created with a compilation, but doesn't expose it.

Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: but you could expose? you already added option so might be okay but adding one more parameter seems more work than just exposing option from compilation data.

**Comment by mavasani** [comment]
Tagging @MattGertz for ask mode approval

**Comment by MattGertz** [comment]
(Tag me again when you have one more signoff.)

**Comment by mavasani** [comment]
@srivatsn Can you take a look?

**Comment by mavasani** [comment]
Tagging @MattGertz again as we have 2 signoffs.

**Comment by mavasani** [comment]
Thanks @MattGertz - merging in.

**Pull request merged**